engines/macventure/text.cpp

```diff
@@ -0,0 +1,200 @@
+* are too numerous to list here. Please refer to the COPYRIGHT
+* file distributed with this source distribution.
+*
+* This program is free software; you can redistribute it and/or
+* modify it under the terms of the GNU General Public License
+* as published by the Free Software Foundation; either version 2
+* of the License, or (at your option) any later version.
+
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+
+* You should have received a copy of the GNU General Public License
+* along with this program; if not, write to the Free Software
+* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*
+*/
+
+#include "macventure/text.h"
+
+namespace MacVenture {
+TextAsset::TextAsset(MacVentureEngine *engine, ObjID objid, ObjID source, ObjID target, Container *container, bool isOld, const HuffmanLists *huffman) {
+	_id = objid;
+	_sourceObj = source;
+	_targetObj = target;
+	_container = container;
+	_huffman = huffman;
+	_isOld = isOld;
+	_engine = engine;
+
+	if (_isOld) {
+		decodeOld();
+	} else {
+		decodeHuffman();
+	}
+}
+
+void TextAsset::decodeOld() {
+	Common::SeekableReadStream *res = _container->getItem(_id);
+	uint16 strLen = res->readUint16BE();
+	Common::BitStream32BELSB stream(res, true);
+	char* str = new char[strLen + 1];
+	bool lowercase = false;
+	char c;
+	for (uint16 i = 0; i < strLen; i++) {
+		char val = stream.getBits(5);
+		if (val == 0x0) { // Space
+			c = ' ';
+		} else if (val >= 0x1 && val <= 0x1A) {
+			if (lowercase) { // Ascii a-z
+				c = val + 0x60;
+			} else { // Ascii A-Z
+				c = val + 0x40;
+			}
+			lowercase = true;
+		} else if (val == 0x1B) {
+			if (lowercase) {
+				c = '.';
+			} else {
+				c = ',';
+			}
+			lowercase = true;
+		} else if (val == 0x1C) {
+			if (lowercase) {
+				c = 0x27; // Ascii '
+			} else {
+				c = 0x22; // Ascii '"'
+			}
+			lowercase = true;
+		} else if (val == 0x1D) { // Composite
+			ObjID subval = stream.getBits(16);
+			Common::String child;
+			if (subval & 0x8000) {
+				// Composite object id
+				subval ^= 0xFFFF;
+				child = getNoun(subval);
+			} else {
+				// Just another id
+				// HACK, see below in getNoun()
+				child = *TextAsset(_engine, subval, _sourceObj, _targetObj, _container, _isOld, _huffman).decode();
+			}
+			if (child.size() > 0) {
+				c = '?'; // HACK Will fix later, should append
+			}
+			lowercase = true;
+		} else if (val == 0x1E) {
+			c = stream.getBits(8);
+			lowercase = true;
+		} else if (val == 0x1F) {
+			lowercase = !lowercase;
+		} else {
+			warning("Unrecognized char in old text %d, pos %d", _id, i);
+		}
+		str[i] = c;
+	}
+
+	str[strLen] = '\0';
+	debugC(3, kMVDebugText, "Decoded %d string (old): %s", _id, str);
+	_decoded = Common::String(str);
+}
+
+void TextAsset::decodeHuffman() {
```

Comment:
I am, and I  tried it. It didn't work when I tried it, and I thought it was because ScummVM's decoder worked with equality (`==`), while this one comapred the symbols by greater-than. Although, as pointed out in another comment, it might have been that there was an invalid value being read from beyond the end of a file.